doc/develguide.rst

```diff
@@ -0,0 +1,656 @@
+
+In the earlier design, wrappers, primitives and objects were all seperate
+concepts.  At the JNI layer these are unified by a common element called 
+jvalue.  A ``jvalue`` is a union of all primitives with the jobject.  The jobject
+can represent anything derived from Java object including the pseudo class
+jstring.  To represent these in a common way all jpype objects that 
+represent an instance of a Java resource have a ``__javavalue__`` field which 
+should be of type ``_jpype.PyJPValue``.  Rather than forcing checking of individual
+types, we use duck typing of simply checking for a ``PyJPValue`` in this field.
+In addition to the union, the jvalue also carries a hidden type.  Thus hidden
+type is used to help in casting the object and resolving method overloads.
+We will discuss this object further in the CPython section.
+
+``jclass``
++++++++++++++++++++++++++++++++
+
+In addition class wrappers have a ``_jpype.PyJPClass`` field which represents the 
+internal class wrapper.  This class wrapper holds the reflection api used to create
+the class methods and fields.  This field is stored as the ``__javaclass__``
+in the class wrapper.  As the class wrapper is used to create an object instance
+``__javaclass__`` also appears in the objects.  Only objects that have a ``__javaclass__``
+and lack a ``__javavalue__`` are treated as class wrappers for the purposes of 
+duck typing.
+
+Because the Java class is both an object and type, we used the duck typing to 
+allow the class pointer to be converted into a class instance.  This is exposed
+as property called ``class_``.  The this ``class_`` property is the equivalent of the 
+.class member of a Java class name or ``getClass()`` of a class instance.  As it 
+is an object instance of ``java.lang.Class`` it can be used for any reflection 
+needs by the user.  However, when working with the jpype core module, on
+needs to be aware of that the this class instance is not available until 
+the key wrappers are created.  See bootstrapping for further details.
+
+``jarray``
++++++++++++++++++++++++++++++++
+
+Java arrays are a special form of objects.  They have no real methods as they
+are not extendable.  To help in accessing the additonal special methods associated 
+with an array, Java array instances have an additional field ``__javaarray__`` of
+type ``PyJParray``.
+
+``jstring``
++++++++++++++++++++++++++++++++
+
+For most practical purposes Java strings are treated as objects.  However, they 
+also need to be able to interact with Python strings.  In the previous version,
+strings were automatically converted to Python strings on return.  This resulted
+in rather strange behavior when interacting with the methods of ``java.lang.String``
+as rather than getting the expected Java object for chaining of commands, the 
+string object would revert to Python.  To avoid this fate, we now require string
+objects to be convert manually with the ``str()`` method in Python.  There are 
```

Comment:
This string handling is an API breakage isn't it? Is it really mandatory?

Reply:
In the previous system is was impossible to keep a string on the java side for any length of time.  Even if you you were just passing around a large string from java to java, it was forced to come back to python.  There was a global switch that was supposed to keep java strings on the java side, but it was filled with bugs.  The revised version is exactly as the switch state was supposed to do.  You call a method and get back a java.lang.String and if you want to pay to convert it call str() on it.

The previous string behavior was very much regrettable.   It was impossible prototype simple string codes because methods things were the user clearly wanted to get back a java string like substring() instead gave pystring instead.  

Mode switching like we had was impossible to maintain for any realistic use  (so that one was out) and our previous default and getting pystring every time a string prohibited a number of patterns.  

Can it be undone, yes (with effort).  In the code returning jvalues back from a method call and in the field string we can add a special case that forces the string to decode back to python object.  As the path string path is no longer internally different than  any other handler (part of the code simplification), we would need to check each of the paths.   

So the question is what will this break.  
- If the user got a string  and expected it to have python methods, it fails with method not found.
- Jstrings convert themselves automatically when going to string concatenation or formatting so no breakage here.
- left equivalence and right equivalence  are not equal.  Jstring knows how to compare to a pystring properly, but pystring does not know how to compare to jstring. 
- Use as dictionary keys, does not work as expected especially if mixed.
    
So as to what you were asking."is it mandatory."  I think so, the previous design left a lot of bad behavior in which ever mode was selected by the global switch.   We can special case it to push this down the road and make old code but it still leaves a lot of holes in the API.  I tested all of my code base with the new behavior and found only two places that needed a str() call added.